crypto/thread/arch/thread_posix.c

```diff
@@ -0,0 +1,305 @@
+{
+    pthread_mutex_t *mutex_p;
+
+    mutex_p = (pthread_mutex_t *)mutex;
+    pthread_mutex_lock(mutex_p);
+}
+
+void ossl_crypto_mutex_unlock(CRYPTO_MUTEX *mutex)
+{
+    pthread_mutex_t *mutex_p;
+
+    mutex_p = (pthread_mutex_t *)mutex;
+    pthread_mutex_unlock(mutex_p);
+}
+
+void ossl_crypto_mutex_free(CRYPTO_MUTEX **mutex)
+{
+    pthread_mutex_t **mutex_p;
+
+    if (mutex == NULL)
+        return;
+
+    mutex_p = (pthread_mutex_t **)mutex;
+    if (*mutex_p != NULL)
+        pthread_mutex_destroy(*mutex_p);
+    OPENSSL_free(*mutex_p);
+    *mutex = NULL;
+}
+
+CRYPTO_CONDVAR *ossl_crypto_condvar_new(void)
+{
+    pthread_cond_t *cv_p;
+
+    if ((cv_p = OPENSSL_zalloc(sizeof(*cv_p))) == NULL)
+        return NULL;
+    if (pthread_cond_init(cv_p, NULL) != 0) {
+        OPENSSL_free(cv_p);
+        return NULL;
+    }
+    return (CRYPTO_CONDVAR *) cv_p;
+}
+
+void ossl_crypto_condvar_wait(CRYPTO_CONDVAR *cv, CRYPTO_MUTEX *mutex)
+{
+    pthread_cond_t *cv_p;
+    pthread_mutex_t *mutex_p;
+
+    cv_p = (pthread_cond_t *)cv;
+    mutex_p = (pthread_mutex_t *)mutex;
+    pthread_cond_wait(cv_p, mutex_p);
+}
+
+void ossl_crypto_condvar_broadcast(CRYPTO_CONDVAR *cv)
+{
+    pthread_cond_t *cv_p;
+
+    cv_p = (pthread_cond_t *)cv;
+    pthread_cond_broadcast(cv_p);
+}
+
+void ossl_crypto_condvar_free(CRYPTO_CONDVAR **cv)
+{
+    pthread_cond_t **cv_p;
+
+    if (cv == NULL)
+        return;
+
+    cv_p = (pthread_cond_t **)cv;
+    if (*cv_p != NULL)
+        pthread_cond_destroy(*cv_p);
+    OPENSSL_free(*cv_p);
+    *cv_p = NULL;
+}
+
+void ossl_crypto_mem_barrier(void)
+{
+# if defined(__clang__) || defined(__GNUC__)
+    __sync_synchronize();
+# elif !defined(OPENSSL_NO_ASM)
+#  if defined(__alpha__) // Alpha
+    __asm__ volatile("mb" : : : "memory");
+#  elif defined(__amd64__) || defined(__i386__) || defined(__i486__) \
+    || defined(__i586__)  || defined(__i686__) || defined(__i386) // x86
+    __asm__ volatile("mfence" : : : "memory");
+#  elif defined(__arm__) || defined(__aarch64__) // ARMv7, ARMv8
+    __asm__ volatile("dmb ish" : : : "memory");
+#  elif defined(__hppa__) // PARISC
+    __asm__ volatile("" : : : "memory");
+#  elif defined(__mips__) // MIPS
+    __asm__ volatile("sync" : : : "memory");
+#  elif defined(__powerpc__) || defined(__powerpc64__) // power, ppc64, ppc64le
+    __asm__ volatile("sync" : : : "memory");
+#  elif defined(__sparc__)
+    __asm__ volatile("ba,pt	%%xcc, 1f\n\t" \
+                     " membar	#Sync\n"   \
+                     "1:\n"                \
+                     : : : "memory");
+#  elif defined(__s390__) || defined(__s390x__) // z
+    __asm__ volatile("bcr 15,0" : : : "memory");
+#  elif defined(__riscv) || defined(__riscv__) // riscv
+    __asm__ volatile("fence iorw,iorw" : : : "memory");
+#  else // others, compiler only
+    __asm__ volatile("" : : : "memory");
+#  endif
+# else
+    // compiler only barrier
+    __asm__ volatile("" : : : "memory");
+# endif
+}
```

Comment:
Okay, for this, bravo... :smile: 

There are some C99 comments which should be changed to C89 though.